pkg/clockcache/cache.go

```diff
@@ -0,0 +1,201 @@
+	// stores indexes into storage
+	elements map[interface{}]*Element
+	storage  []Element
+
+	// CLOCK sweep state, must have write lock
+	next int
+}
+
+type Element struct {
+	// The value stored with this element.
+	key   interface{}
+	Value interface{}
+
+	//CLOCK marker if this is recently used
+	used uint32
+
+	// pad Elements out to be cache aligned
+	_padding [24]byte
+}
+
+func WithMax(max uint64) *Cache {
+	if max < 1 {
+		panic("must have max greater than 0")
+	}
+	return &Cache{
+		elements: make(map[interface{}]*Element, max),
+		storage:  make([]Element, 0, max),
+	}
+}
+
+// Insert a key/value mapping into the cache if the key is not already present
+// returns the value present in the map, and true if it is newley inserted
+func (self *Cache) Insert(key interface{}, value interface{}) (canonicalValue interface{}, inserted bool) {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	_, canonicalValue, inserted = self.insert(key, value)
+	return
+}
+
+// Insert a bactch of keys with their corresponding values.
+// This function will _overwrite_ the keys and values slices with their
+// canonical versions.
+func (self *Cache) InsertBatch(keys []interface{}, values []interface{}) {
+	if len(keys) != len(values) {
+		panic(fmt.Sprintf("keys and values are not the same len. %d keys, %d values", len(keys), len(values)))
+	}
+	values = values[:len(keys)]
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	for idx := range keys {
+		keys[idx], values[idx], _ = self.insert(keys[idx], values[idx])
+	}
+	return
+}
+
+func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, inserted bool) {
+	elem, present := self.elements[key]
+	if present {
+		return elem.key, elem.Value, false
+	}
+
+	var insertLocation *Element
+	if len(self.storage) >= cap(self.storage) {
+		insertLocation = self.evict()
+		*insertLocation = Element{key: key, Value: value}
+	} else {
+		self.storage = append(self.storage, Element{key: key, Value: value})
+		insertLocation = &self.storage[len(self.storage)-1]
+	}
+
+	self.elements[key] = insertLocation
+	return key, value, true
+}
+
+func (self *Cache) evict() (insertPtr *Element) {
+	for {
+		insertLocation, evicted := self.tryEvict()
+		if evicted {
+			return insertLocation
+		}
+	}
+}
+
+func (self *Cache) tryEvict() (insertPtr *Element, evicted bool) {
+	if self.next >= len(self.storage) {
+		self.next = 0
+	}
+
+	evicted = false
+	reachedEnd := false
+	for !evicted && !reachedEnd {
+		elem := &self.storage[self.next]
+		if elem.used != 0 {
+			elem.used = 0
+		} else {
+			insertPtr = elem
+			key := elem.key
+			delete(self.elements, key)
+			evicted = true
+		}
+		self.next += 1
+		reachedEnd = self.next >= len(self.storage)
+	}
+
+	return
+}
+
+// tries to get a batch of keys and store the corresponding values is valuesOut
+// returns the number of keys that were actually found.
+// NOTE: this function does _not_ preserve the order of keys; the first numFound
```

Comment:
The real reason is because the API was designed around non-pointer keys. However, the current version has the advantage of constant-time discovery of hits/misses. In our labels cache, the expected use looks something like:
```go
numFound := labelsCache.GetValue(labelIds, labels)
if numFound < len(labelIds) {
    missed := fetchLabels(labelIds[numFound:])
    copy(labels[numFound:], missed)
}
sort.Slice(labels, ...)
return labels
```
which I think lends itself better to the current API